util/eventbus/debug.go

```diff
@@ -9,6 +9,147 @@ import (
 	"sync/atomic"
 )
 
+// A Debugger offers access to a bus's privileged introspection and
+// debugging facilities.
+//
+// The debugger's functionality is intended for humans and their tools
+// to examine and troubleshoot bus clients, and should not be used in
+// normal codepaths.
+//
+// In particular, the debugger provides access to information that is
+// deliberately withheld from bus clients to encourage more robust and
+// maintainable code - for example, the sender of an event, or the
+// event streams of other clients. Please don't use the debugger to
+// circumvent these restrictions for purposes other than debugging.
+type Debugger struct {
+	bus *Bus
+}
+
+// Clients returns a list of all clients attached to the bus.
+func (d *Debugger) Clients() []*Client {
+	return d.bus.listClients()
+}
+
+// PublishQueue returns the contents of the publish queue.
+//
+// The publish queue contains events that have been accepted by the
+// bus from Publish() calls, but have not yet been routed to relevant
+// subscribers.
+//
+// This queue is expected to be almost empty in normal operation. A
+// full publish queue indicates that a slow subscriber downstream is
+// causing backpressure and stalling the bus.
+func (d *Debugger) PublishQueue() []PublishedEvent {
+	return d.bus.snapshotPublishQueue()
+}
+
+func (d *Debugger) SubscribeQueue(client *Client) []DeliveredEvent {
+	return client.snapshotSubscribeQueue()
+}
+
+// WatchBusSync calls fn for every event passing through the bus.
+//
+// Calls to fn are serialized, and events are delivered in the bus's
+// global publication order (see "Concurrency properties" in the
+// package docs).
+//
+// fn is called after a published event has been matched to its
+// subscribers, but before the event is queued for delivery. fn blocks
+// delivery of the event until it returns, and therefore also blocks
+// the processing of all subsequent events.
+//
+// As such, this synchronous interface should be used with care, as
+// doing excessive work within fn will degrade the performance of the
+// entire bus.
+//
+// WatchBusSync returns a cancellation function that stops delivering
+// events to fn when called.
+func (d *Debugger) WatchBusSync(fn func(RoutedEvent)) (cancel func()) {
```

Comment:
Given the existence of the non-sync versions of these, do we need to (yet) export the sync versions? (Maybe you have a use case in mind, in which case it's fine, I just wonder if we can avoid it for now)

Reply:
Unsure. The Sync methods have one major useful property, which is that you get happens-before edges between the watch funcs and bus lifecycle points. I figured that might be useful in tests, so that test watchers can effortlessly synchronize themselves "after publication, but before delivery". But, I don't know if that would be useful yet.

I've unexported the Sync methods for now, but I'm keeping their code and docstrings in good shape for re-publishing easily, if we discover we want them. We can easily delete them later on, but then at least they'll be in the git history if we find we want them.

Reply:
Discussed offline: I'm deleting the *Sync methods for now, as they're easy enough to add back and we couldn't come up with a _definite_ need for their synchronous semantics anywhere. The async methods are easier to use correctly, so we'll stick with those until we find a reason not to.